pkg/storage/scrub.go

```diff
@@ -92,84 +88,171 @@ func CheckRepo(ctx context.Context, imageName string, imgStore storageTypes.Imag
 		return results, ctx.Err()
 	}
 
-	dir := path.Join(imgStore.RootDir(), imageName)
-	if !imgStore.DirExists(dir) {
-		return results, errors.ErrRepoNotFound
-	}
+	var lockLatency time.Time
+
+	imgStore.RLock(&lockLatency)
+	defer imgStore.RUnlock(&lockLatency)
 
-	oci, err := umoci.OpenLayout(dir)
+	// check image structure / layout
+	ok, err := imgStore.ValidateRepo(imageName)
 	if err != nil {
 		return results, err
 	}
 
-	defer oci.Close()
-
-	var lockLatency time.Time
-
-	imgStore.RLock(&lockLatency)
-	defer imgStore.RUnlock(&lockLatency)
+	if !ok {
+		return results, errors.ErrRepoBadLayout
+	}
 
-	buf, err := os.ReadFile(path.Join(dir, "index.json"))
+	// check "index.json" content
+	indexContent, err := imgStore.GetIndexContent(imageName)
 	if err != nil {
 		return results, err
 	}
 
 	var index ispec.Index
-	if err := json.Unmarshal(buf, &index); err != nil {
+	if err := json.Unmarshal(indexContent, &index); err != nil {
 		return results, errors.ErrRepoNotFound
 	}
 
-	listOfManifests := []ispec.Descriptor{}
+	listOfManifests := make(map[godigest.Digest]ScrubImageResult)
 
 	for _, manifest := range index.Manifests {
-		if manifest.MediaType == ispec.MediaTypeImageIndex {
-			buf, err := os.ReadFile(path.Join(dir, "blobs", manifest.Digest.Algorithm().String(), manifest.Digest.Encoded()))
-			if err != nil {
-				tagName := manifest.Annotations[ispec.AnnotationRefName]
-				imgRes := getResult(imageName, tagName, errors.ErrBadBlobDigest)
-				results = append(results, imgRes)
-
-				continue
-			}
+		tag := manifest.Annotations[ispec.AnnotationRefName]
+		imageResult := getManifestResult(ctx, manifest, imgStore, imageName, tag, listOfManifests)
+		results = append(results, imageResult)
+	}
+
+	return results, nil
+}
+
+func getManifestResult(
+	ctx context.Context, manifest ispec.Descriptor, imgStore storageTypes.ImageStore, imageName, tag string,
+	listOfManifests map[godigest.Digest]ScrubImageResult,
+) ScrubImageResult {
```

Comment:
Why do we need to return the results, if we add them to the map anyway?

Reply:
Fixed. Now this function is called `scrubManifest` and it just adds the results to the map without returning them.